### PR TITLE
Made it so that only the most confident result is voiced when you cli…

### PR DESCRIPTION
**Pull request by marjasi**
…ck CAPTURE.

### CODE REVIEW

**Comment by EdvardasDlugauskas** [comment]
[[LINQ] Make the WF app say only the most confident recognition result](https://trello.com/c/n8SwrqDB/24-linq-make-the-wf-app-say-only-the-most-confident-recognition-result)

File: ShopLens/ShopLensForms/ShopLens.cs [hunk 1]
```diff
@@ -57,21 +57,26 @@ private void CommandRecognized_WhatIsThis(object sender, EventArgs e)
         {
             if (live_video.Image != null)
             {
-                Bitmap image = (Bitmap)live_video.Image.Clone();
-                capture_picture.Image = (Bitmap)live_video.Image.Clone();
+                var image = (Image)live_video.Image.Clone();
```
Comment (EdvardasDlugauskas) [inline ShopLens/ShopLensForms/ShopLens.cs]: I thought we agreed to comment out the line `capture_picture.Image = image`? Did changing Bitmap to Image fix that crash?

File: ShopLens/ShopLensForms/ShopLens.cs [hunk 2]
```diff
@@ -57,21 +57,26 @@ private void CommandRecognized_WhatIsThis(object sender, EventArgs e)
...
+                capture_picture.Image = image;
 
                 var ms = new MemoryStream();
                 image.Save(ms, System.Drawing.Imaging.ImageFormat.Jpeg);
 
                 var classificationResults = _imageClassifying.ClassifyImage(ms.ToArray());
 
                 var resultStrings = classificationResults.Select(pair => $"{pair.Key} - {(int)(pair.Value * 100)} percent.");
```
Comment (EdvardasDlugauskas) [inline ShopLens/ShopLensForms/ShopLens.cs]: I think `resultStrings` is not needed anymore

File: ShopLens/ShopLensForms/ShopLens.cs [hunk 3]
```diff
@@ -57,21 +57,26 @@ private void CommandRecognized_WhatIsThis(object sender, EventArgs e)
...
-                }
+                _textVoicer.SayMessage("This is");
+
+                //Order by probability values and take the first label name.
+                string mostConfidentResult = classificationResults.OrderByDescending(x => x.Value).FirstOrDefault().ToString();
+
+                //Format string so only the name of the product is said.
+                string[] results = mostConfidentResult.Split(',');
+
+                _textVoicer.SayMessage(results[0]);
```
Comment (EdvardasDlugauskas) [inline ShopLens/ShopLensForms/ShopLens.cs]: That's a very interesting way of doing things 😄 

How about:
```
// Since the results look like <'tomato', 0.99>, we can just take the Key
var prediction = classificationResults.OrderByDescending(x => x.Value).FirstOrDefault().Key;
if (prediction == null)
{
    // don't forget to handle null !
}
else
 {
    _textVoicer.SayMessage(prediction);
}
```

File: ShopLens/ShopLensForms/ShopLens.cs [hunk 4]
```diff
@@ -127,7 +132,33 @@ private void PAUSE_Click(object sender, EventArgs e)
 
         private void CAPTURE_Click(object sender, EventArgs e)
         {
+            if (live_video.Image != null)
```
Comment (EdvardasDlugauskas) [inline ShopLens/ShopLensForms/ShopLens.cs:121]: Wait a second 🕐 ... Is this method just copy-paste 🤔 🤔 🤔 

Remember, we want to be DRY 🏜 🐪 (Don't Repeat Yourself - OOP principle) 

But, in all seriousness, this method is just here for testing purposes and will be removed? Or will we still have it for testing purposes?

Either way, I don't think we should merge two copy-pasted methods in master, not because it's very bad, but because code hygiene. So we should either throw it out, or make it DRY.

**Review by EdvardasDlugauskas — CHANGES_REQUESTED** [review]
Hey it's all good.

Unfortunately now there's 2 methods which are basically the same. You should do something about them and then I think we'll be good to merge 👍

File: ShopLens/ShopLensForms/ShopLens.cs (hunk 1 above)
Comment (marjasi) [reply ShopLens/ShopLensForms/ShopLens.cs]: No it didn't fix the crash, it's just that I made this branch from my voice recognizer interface branch when I was trying to fix the thread issue and I kinda left it there uncommented..

File: ShopLens/ShopLensForms/ShopLens.cs (hunk 2 above)
Comment (marjasi) [reply ShopLens/ShopLensForms/ShopLens.cs]: Yeah, it's not.

File: ShopLens/ShopLensForms/ShopLens.cs (hunk 3 above)
Comment (marjasi) [reply ShopLens/ShopLensForms/ShopLens.cs]: >>compiler tells you this ain't a string
>>you write .ToString to it
>>compiler doesn't complain no more, but code is sloppy
>>??!!??
>>PROFIT!
>>then you figure out how to make it so that only the name of the product is said
>>you split the string
>>it says the name of the product
>>and in the process of all of this I didn't get that I can just get the key

File: ShopLens/ShopLensForms/ShopLens.cs (hunk 4 above)
Comment (marjasi) [reply ShopLens/ShopLensForms/ShopLens.cs:121]: Yeah, you're right, I'll make it so that the what is this command event method runs this method. That will probably make it 🌵 🐫 🏜  DRY  🏜 🐫 🌵 enough.

File: ShopLens/ShopLensForms/ShopLens.cs [hunk 5]
```diff
@@ -134,26 +109,32 @@ private void CAPTURE_Click(object sender, EventArgs e)
...
+                if (mostConfidentResult == null) 
+                {
+                    _textVoicer.SayMessage(noLblError);
+                }
+                else
+                {
+                    _textVoicer.SayMessage(mostConfidentResult);
+                }
+
+                Thread.Sleep(500);  //Why is this here?
```
Comment (marjasi) [inline ShopLens/ShopLensForms/ShopLens.cs]: Why is this here? Why do we need to stop the main Thread? What happens when you delete this?

Comment (marjasi) [reply ShopLens/ShopLensForms/ShopLens.cs]: I noticed that if you CAPTURE too many times in a short span of time the program throws an exception related to the "Thread Bug".

Comment (marjasi) [reply ShopLens/ShopLensForms/ShopLens.cs]: But this only happens when using the voice command "what is this" -  you can click the CAPTURE button as many times as you want.

Comment (EdvardasDlugauskas) [reply ShopLens/ShopLensForms/ShopLens.cs]: Good point, should delete this line

**Review by EdvardasDlugauskas — APPROVED** [review]
